### DISCUSSION TITLE
Collections spread operator in combination with params

### DISCUSSION THREAD

**Discussion by glucaci** [dotnet/csharplang#3648] [Language Ideas]
The spreading operator will be very useful when you want to pass into params multiple items and a list without concatenation before.

Current:
```csharp
public void StartProcess()
{
    var predefined = new [] { 1, 2, 3 };

    Process(predefined.Concat(new [] { 4, 5 }).ToArray());
}

private void Process(params int[] numbers)
{
    // Do work
}
```

Proposal:
```csharp
public void StartProcess()
{
    var predefined = new [] { 1, 2, 3 };

    Process([...predefined], 4, 5); // [...<collection>] or something similar
}

private void Process(params int[] numbers)
{
    // Do work
}
```

**Comment by huoyaoyuan** [comment]
It's quite hard to let the operator to work *like this way*. Unlike many script languages that has only 1 type used both as array and list, there are many types of collections in C#, and there's no unified concept of "concat".
However, we do have unified collection initialization expression, which can be a target for collection spread operator:
```
new List<int>{ [..predefined], 4, 5}
```
But there are still open questions:
CLR arrays are fixed size. How to create it with correct length? There's no guarantee that the spreaded collection implements `Length` correctly.

**Reply by spixy** [reply]
Then we will get `IndexOutOfRangeException` which is fine. Then it will be problem of `Length` implementation, not problem of the language or CLR.

**Comment by theunrepentantgeek** [comment]
I'm not sure that the surrounding `[` `]` are needed, the syntax could be just
```
Process(...predefined, 4, 5);
```

In alignment with other features of C# that work with _patterns_ (not to be confused with _pattern matching_), the syntax could be supported by any collection type that has an `AddRange()` method (instance or extension) with an appropriate signature.

For an array type
```
private void Process(params int[] numbers)
```
Someone would need to write an `AddRange()` method, perhaps one that returns a new array.

For a List type
```
private void Process(params List<int> numbers)
```
it could be supported "out of the box".

Brainstorming aside, though, I'm not sure that this adds enough value to the language to justify the investment.

If we ever got `params IEnumerable<int>` (which has been suggested multiple times #178, #2141) we'd be able to write
```
Process(predefined.Concat(4, 5));
```
and get the same result.
